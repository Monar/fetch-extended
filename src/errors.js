export class TimeoutError extends Error {
  constructor(msg, request) {
    super(msg);
    this.request = request;
  }
}

