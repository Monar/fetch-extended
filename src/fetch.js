import {
  mergeHeaders,
  constructSearchParams,
  timeoutPromise,
} from './utils';

export const fetchxDefaults = {
  headers: {},
  options: {
    timeout: 60000,
  },
};

export function fetchx(...args) {
  return fetchWrapper(fetch, fetchxDefaults, ...args);
}

export function getFetchx(defaultHeaders = {}, defautlOptions = {}) {
  const defaults = {
    headers: Object.assign({}, fetchxDefaults.headers, defaultHeaders),
    options: Object.assign({}, fetchxDefaults.options, defautlOptions),
  };

  return (...args) => fetchWrapper(fetch, defaults, ...args);
}

export function fetchWrapper(fetch, defaults, url, options = {}) {
  const headers = mergeHeaders(defaults.headers, options.headers);
  const searchParams = constructSearchParams(options.query);
  const fetchURL = searchParams ? `${url}?${searchParams}` : url;
  const fetchOptions = Object.assign({}, defaults.options, options, { headers });

  const promise = fetch(fetchURL, fetchOptions);
  const { timeout } = fetchOptions;
  if (typeof timeout === 'number') {
    return timeoutPromise(timeout, promise, { url: fetchURL, options: fetchOptions });
  }

  return promise;
}
