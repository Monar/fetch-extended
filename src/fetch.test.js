/* eslint-env jest */
import 'isomorphic-fetch';

import { fetchWrapper, fetchxDefaults } from './fetch';
import { TimeoutError } from './errors.js'

describe('Fetch implementation', () => {
  it('should work with just an url', () => {
    const fetch = url => Promise.resolve(url);
    const url = 'http://test.address.com';

    const promise = fetchWrapper(fetch, fetchxDefaults, url);

    expect(promise).resolves.toBe(url);
  });

  it('should work with search query', () => {
    const fetch = url => Promise.resolve(url);
    const url = 'http://test.address.com';
    const query = {
      miłość: 'Love&Cuddle',
      options: [1, 2, null, 'tree'],
      'ups not again': undefined,
    };

    const expected = {
      miłość: 'Love&Cuddle',
      options: '1,2,,tree',
      'ups not again': 'undefined',
    };

    const promise = fetchWrapper(fetch, fetchxDefaults, url, { query });
    promise.then(url => {
      const test = new URL(url);
      Object.keys(query).forEach(key => {
        expect(test.searchParams.get(key)).toBe(expected[key]);
      });
    });
  });

  it('should all the options', () => {
    const fetch = (url, options) => Promise.resolve(options);
    const url = 'http://test.address.com';
    const testHeaders = new Headers();
    testHeaders.append('session-id', 'session-key');

    const testOptions = {
      method: 'GET',
      headers: testHeaders,
      mode: 'cors',
      cache: 'default',
    };

    const promise = fetchWrapper(fetch, fetchxDefaults, url, testOptions);
    expect(promise).resolves.toMatchSnapshot();
  });

  it('should reject when reach timeout', async () => {
    const fetch = () =>
      new Promise(resolve => {
        setTimeout(() => resolve('lol'), 1000);
      });

    const url = 'http://www.fuu.com/';
    const promise = fetchWrapper(fetch, fetchxDefaults, url, { timeout: 0 });
    await expect(promise).rejects.toBeInstanceOf(TimeoutError);
  });

  it('should work with custom queryParser', async () => {
    const url = 'http://test.address.com';
    const fetch = url => Promise.resolve(url);
    const testOptions = {
      queryParser,
      query: { foo: ['uno', 'dos,...'] },
    };

    const promise = fetchWrapper(fetch, fetchxDefaults, url, testOptions);
    expect(promise).resolves.toMatchSnapshot();

    function queryParser(query = {}) {
      const searchParams = new URLSearchParams();
      Object.keys(query).forEach(key => { 
          if(Array.isArray(query[key])) {
            query[key].forEach(value => searchParams.append(key, value));
            return
          }
          searchParams.append(key, query[key])
        });
      return searchParams.toString();
    }
  });
});
