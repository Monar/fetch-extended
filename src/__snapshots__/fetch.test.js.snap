// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch implementation should all the options 1`] = `
Object {
  "cache": "default",
  "headers": Headers {
    "_headers": Object {
      "session-id": Array [
        "session-key",
      ],
    },
  },
  "method": "GET",
  "mode": "cors",
  "timeout": 60000,
}
`;
